src/lib/update-lock.ts

```diff
@@ -96,69 +88,94 @@ function dispose(
-				return writeLock(appId)
-					.tap((release: () => void) => {
-						const lockDir = getPathOnHost(lockPath(appId));
-						return Bluebird.resolve(fs.readdir(lockDir))
-							.catchReturn(ENOENT, [])
-							.mapSeries((serviceName) => {
-								return Bluebird.mapSeries(
-									lockFilesOnHost(appId, serviceName),
-									(tmpLockName) => {
-										return (
-											Bluebird.try(() => {
-												if (force || lockOverride) {
-													return lockfile.unlock(tmpLockName);
-												}
-											})
-												.then(() => {
-													return lockfile.lock(tmpLockName, LOCKFILE_UID);
-												})
-												// If lockfile exists, throw a user-friendly error.
-												// Otherwise throw the error as-is.
-												// This will interrupt the call to Bluebird.using, so
-												// dispose needs to be called even though it's referenced
-												// by .disposer later.
-												.catch((error) => {
-													return dispose(release, appId).throw(
-														lockfile.LockfileExistsError.is(error)
-															? new UpdatesLockedError(
-																	`Lockfile exists for ${JSON.stringify({
-																		serviceName,
-																		appId,
-																	})}`,
-															  )
-															: (error as Error),
-													);
-												})
-										);
-									},
-								);
-							});
-					})
-					.disposer((release: () => void) => dispose(release, appId));
-			})
-			.catch((err) => {
-				throw new InternalInconsistencyError(
-					`Error getting lockOverride config value: ${err?.message ?? err}`,
-				);
-			});
-	};
+	// Sort appIds so they are always locked in the same sequence
+	const sortedIds = appIdsToLock.sort();
 
-	const disposer = takeTheLock();
+	let lockOverride: boolean;
+	try {
+		lockOverride = await config.get('lockOverride');
+	} catch (err) {
+		throw new InternalInconsistencyError(
+			`Error getting lockOverride config value: ${err?.message ?? err}`,
+		);
+	}
 
-	return Bluebird.using(disposer, fn as () => PromiseLike<T>);
+	const releases: Array<() => void> = [];
+	try {
+		for await (const id of sortedIds) {
+			const lockDir = getPathOnHost(lockPath(id));
+			// Acquire write lock for appId
+			releases.push(await writeLock(id));
+			// Get list of service folders in lock directory
+			let serviceFolders: string[] = [];
+			try {
+				serviceFolders = await fs.readdir(lockDir);
+			} catch (err) {
+				if (ENOENT(err)) {
+					// Default to empty list of folders
+					serviceFolders = [];
+				} else {
+					throw err;
+				}
+			}
+			// Attempt to create a lock for each service
+			for await (const service of serviceFolders) {
+				await lockService(id, service, { force, lockOverride });
+			}
+		}
+
+		// Resolve the function passed
+		return Promise.resolve(fn());
+	} finally {
+		// Cleanup locks
+		for (const release of releases) {
+			release();
+		}
+		for await (const lockedApp of sortedIds) {
+			dispose(lockedApp);
+		}
+	}
```

Comment:
Since `release` & `dispose` will run the same number of times, i.e. once for each `appId`, why not deduplicate here by calling `release` in `dispose`? As far as semantics goes, `release`-ing the writeLock is part of `disposing` of the lock effects, so incorporating `release` into `dispose` makes sense. (I believe that's how it was done before too)

Reply:
yeah I wanted to think of how to clean this up for sure. It's weird though because you can't assume you'll have the same number of releases and applications to unlock. What if you never lock any services but you locked an application ? I know the cost is nothing (I think) but it feels wrong.

tl;dr I did plan to come back to this.

Reply:
To clarify, when you're talking about `locking applications`, you're referring to `rwlock`'s `writeLock`? But when you're talking about `locking services`, you're referring to `lockfile`?

So we know for sure that:

- Whenever we pass an `appId` to `writeLock`, that appId does not currently have a write lock on it, per `rwlock`'s stipulation that "there may be only one writer at a time". So every `appId` will definitely have a `release` function associated with it which needs to be called.
- Whenever we try to `unlock` then `lock` each app's services using `lockfile`, 0-1 locks for each service will be created. So 0-n locks will be created for any given appId, where n = number of services.

So, you're right that you could end up never locking any services, but locking an application. However, this is the only condition that's possible - you'll never be able to lock services but not lock an application.

Given the above, we know that if we want to combine the `lockfile.unlock` and the `release` part of `dispose`, we need to 
- always release the `writeLock` for an appId
- handle 0-n `lockfile.unlock`s for an appId's services, where n = number of services

The current implementation of the [lockfile.unlock](https://github.com/balena-os/balena-supervisor/pull/1953/files#diff-376068891dd599a79a4b6bf84d850688e14f1d5ebaf089f88518f08e40fc6f4bR75-R78) part of `dispose` is already capable of handling 0-n `lockfile.unlock`s for a service. Therefore we should be able to add `release` to this without anything erroring..

When you say that combining these two "feels wrong", I guess you mean that the `dispose` method originally combined the dispose effects from two different locking mechanisms: `lockfile` and `rwlock`, and that perhaps combining these 2 mechanisms in this way isn't a clear separate of concerns. If that's the case, how about separating them more distinctly?
```
async function dispose(appIdentifier: string | number, release: () => void): Promise<void> {
    const disposeServiceLockfiles = async () => {
        const locks = lockfile.getLocksTaken((p: string) =>
	    p.includes(`${BASE_LOCK_DIR}/${appIdentifier}`),
        );
        await Promise.all(locks.map((l) => lockfile.unlock(l)));
    }

    const disposeAppWriteLock = () => release();

    return await Promise.all([ disposeServiceLockfiles(), disposeAppWriteLock() ]);
}
```

Reply:
I think the way I've done it is pretty nice

Reply:
Looks good in the newest version (https://github.com/balena-os/balena-supervisor/commit/195792a8c1f652b7c860b836b91fbb8301d653c1)!